Place IEDs within a Private element in the Substation section

danyill opened the pull request:
Very much still WIP

Closes #92

danyill committed:
feat: Position IEDs within menu and in Private elements, closes #92

jarradraumati commented:
I go to place an IED I imported and it does not stay, nor is anything written to the doc.

SSD file to reproduce: 
[sample.zip](https://github.com/user-attachments/files/17323920/sample.zip)

danyill commented:
> I go to place an IED I imported and it does not stay, nor is anything written to the doc.
> 
> SSD file to reproduce: [sample.zip](https://github.com/user-attachments/files/17323920/sample.zip)

Fixed, sorry about that

jarradraumati commented:
Yes that has fixed it. Now a bigger issue:

There is quite a severe lag that occurs as the number of IEDs increases. I have no idea what is causing it.

Example file below to replicate:

[XAT_P1_P2_new_IED_designer_lag.zip](https://github.com/user-attachments/files/17326138/XAT_P1_P2_new_IED_designer_lag.zip)

danyill commented:
Thanks for testing @jarradraumati 

The cause is the `querySelector` being called for each render for the `Text` element and its coordinates. I had some good discussion with @ca-d about this (gracias). For now I'll drop support for the `Text` element within IEDs which will clean up this PR rather a lot and then perhaps we can do a re-test.

danyill commented:
I've removed support for the `Text` element and querySelectors which were reducing performance.

I've added a new ability to remove mismatched IEDName elements within the IEDs menu.

![image](https://github.com/user-attachments/assets/5a70d31e-1a56-47d0-bf73-85d3b2683565)

I think the performance is now similar to for other elements. Here's my large-ish test file.

[XAT_Subsystem_Testing.zip](https://github.com/user-attachments/files/17336473/XAT_Subsystem_Testing.zip)

I would be grateful for some testing. There are still a few undo/redo issues which perhaps can be addressed during a code review. Some of the code still reflects design decisions from when the Text elements were there.

jarradraumati commented:
The performance is now comparable with previous versions.

The undo is somewhat broken as you mentioned. I've haven't taken a look at the code. I end up with duplicates.

![image](https://github.com/user-attachments/assets/ddec98ac-3056-41f5-b352-04a96e57381a)

```xml
<Private type="OpenSCD-Linked-IEDs">
	<esld:IEDName esld:name="XAT_T2_C2" esld:x="13" esld:y="7" esld:lx="14" esld:ly="8"/>
	<esld:IEDName esld:name="XAT_T2_C1" esld:x="42" esld:y="3" esld:lx="43" esld:ly="4"/>
	<esld:IEDName esld:name="XAT_T1_P1" esld:x="29" esld:y="14" esld:lx="30" esld:ly="15"/>
	<esld:IEDName esld:name="XAT_T1_C1" esld:x="19" esld:y="15" esld:lx="20" esld:ly="16"/>
	<esld:IEDName esld:name="XAT_T1_C2" esld:x="37" esld:y="17" esld:lx="38" esld:ly="18"/>
	<esld:IEDName esld:name="XAT_CTRL_1" esld:x="13" esld:y="14" esld:lx="14" esld:ly="15"/>
	<esld:IEDName esld:name="XAT_PLC_1" esld:x="27" esld:y="7" esld:lx="28" esld:ly="8"/>
	<esld:IEDName esld:name="XAT_BusB_P2" esld:x="45" esld:y="7" esld:lx="46" esld:ly="8"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="29" esld:y="2" esld:lx="30" esld:ly="3"/>
	<esld:IEDName esld:name="XAT_BusB_P1" esld:x="26" esld:y="18" esld:lx="27" esld:ly="19"/>
	<esld:IEDName esld:name="XAT_BusA_P2" esld:x="33" esld:y="16" esld:lx="34" esld:ly="17"/>
	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="22" esld:y="5" esld:lx="23" esld:ly="6"/>
	<esld:IEDName esld:name="XAT_302_P2" esld:x="11" esld:y="12" esld:lx="12" esld:ly="13"/>
	<esld:IEDName esld:name="XAT_302_P1" esld:x="37" esld:y="5" esld:lx="38" esld:ly="6"/>
	<esld:IEDName esld:name="XAT_302_MU1" esld:x="37" esld:y="29" esld:lx="38" esld:ly="30"/>
	<esld:IEDName esld:name="XAT_302_C1" esld:x="25" esld:y="29" esld:lx="26" esld:ly="30"/>
	<esld:IEDName esld:name="XAT_288_MU1" esld:x="16" esld:y="29" esld:lx="17" esld:ly="30"/>
	<esld:IEDName esld:name="XAT_242_MU2" esld:x="19" esld:y="22" esld:lx="20" esld:ly="23"/>
	<esld:IEDName esld:name="XAT_242_MU1" esld:x="28" esld:y="22" esld:lx="29" esld:ly="23"/>
	<esld:IEDName esld:name="XAT_242_C1" esld:x="40" esld:y="21" esld:lx="41" esld:ly="22"/>
	<esld:IEDName esld:name="XAT_232_P2" esld:x="1" esld:y="15" esld:lx="2" esld:ly="16"/>
	<esld:IEDName esld:name="XAT_232_C1" esld:x="47" esld:y="12" esld:lx="48" esld:ly="13"/>
	<esld:IEDName esld:name="XAT_2242_MU2" esld:x="1" esld:y="2" esld:lx="2" esld:ly="3"/>
	<esld:IEDName esld:name="XAT_2242_MU1" esld:x="61" esld:y="2" esld:lx="62" esld:ly="3"/>
	<esld:IEDName esld:name="XAT_2152_MU2" esld:x="61" esld:y="14" esld:lx="62" esld:ly="15"/>
	<esld:IEDName esld:name="XAT_2152_MU1" esld:x="61" esld:y="6" esld:lx="62" esld:ly="7"/>
	<esld:IEDName esld:name="XAT_T2_P2" esld:x="45" esld:y="16" esld:lx="46" esld:ly="17"/>
	<esld:IEDName esld:name="XAT_T1_P2" esld:x="30" esld:y="5" esld:lx="31" esld:ly="6"/>
	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="24" esld:y="4" esld:lx="25" esld:ly="5"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="27" esld:y="2" esld:lx="29.5" esld:ly="5"/>
	<esld:IEDName esld:name="XAT_242_C1" esld:x="42" esld:y="19" esld:lx="43" esld:ly="20"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="22" esld:y="2" esld:lx="20.5" esld:ly="2"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="27" esld:y="2" esld:lx="25.5" esld:ly="2"/>
	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="23" esld:y="6" esld:lx="22" esld:ly="6"/>
	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="24" esld:y="4" esld:lx="25" esld:ly="5"/>
	<esld:IEDName esld:name="XAT_T1_P2" esld:x="31" esld:y="5" esld:lx="32" esld:ly="6"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="40" esld:y="1" esld:lx="49" esld:ly="2.5"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="35" esld:y="3" esld:lx="36" esld:ly="4"/>
	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="23" esld:y="3" esld:lx="24" esld:ly="4"/>
	<esld:IEDName esld:name="XAT_T2_P1" esld:x="14" esld:y="2" esld:lx="15" esld:ly="3"/>
</Private>
```

Any chance of having an edit IED menu added?

danyill commented:
There should be edit IED if you right-click on the IED or middle click the
text to open the wizard, no?

On Wed, 16 Oct 2024, 20:18 Jarrad Raumati, ***@***.***> wrote:

> The performance is now comparable with previous versions.
>
> The undo is somewhat broken as you mentioned. I've haven't taken a look at
> the code. I end up with duplicates.
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/ddec98ac-3056-41f5-b352-04a96e57381a>
>
> <Private type="OpenSCD-Linked-IEDs">
> 	<esld:IEDName esld:name="XAT_T2_C2" esld:x="13" esld:y="7" esld:lx="14" esld:ly="8"/>
> 	<esld:IEDName esld:name="XAT_T2_C1" esld:x="42" esld:y="3" esld:lx="43" esld:ly="4"/>
> 	<esld:IEDName esld:name="XAT_T1_P1" esld:x="29" esld:y="14" esld:lx="30" esld:ly="15"/>
> 	<esld:IEDName esld:name="XAT_T1_C1" esld:x="19" esld:y="15" esld:lx="20" esld:ly="16"/>
> 	<esld:IEDName esld:name="XAT_T1_C2" esld:x="37" esld:y="17" esld:lx="38" esld:ly="18"/>
> 	<esld:IEDName esld:name="XAT_CTRL_1" esld:x="13" esld:y="14" esld:lx="14" esld:ly="15"/>
> 	<esld:IEDName esld:name="XAT_PLC_1" esld:x="27" esld:y="7" esld:lx="28" esld:ly="8"/>
> 	<esld:IEDName esld:name="XAT_BusB_P2" esld:x="45" esld:y="7" esld:lx="46" esld:ly="8"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="29" esld:y="2" esld:lx="30" esld:ly="3"/>
> 	<esld:IEDName esld:name="XAT_BusB_P1" esld:x="26" esld:y="18" esld:lx="27" esld:ly="19"/>
> 	<esld:IEDName esld:name="XAT_BusA_P2" esld:x="33" esld:y="16" esld:lx="34" esld:ly="17"/>
> 	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="22" esld:y="5" esld:lx="23" esld:ly="6"/>
> 	<esld:IEDName esld:name="XAT_302_P2" esld:x="11" esld:y="12" esld:lx="12" esld:ly="13"/>
> 	<esld:IEDName esld:name="XAT_302_P1" esld:x="37" esld:y="5" esld:lx="38" esld:ly="6"/>
> 	<esld:IEDName esld:name="XAT_302_MU1" esld:x="37" esld:y="29" esld:lx="38" esld:ly="30"/>
> 	<esld:IEDName esld:name="XAT_302_C1" esld:x="25" esld:y="29" esld:lx="26" esld:ly="30"/>
> 	<esld:IEDName esld:name="XAT_288_MU1" esld:x="16" esld:y="29" esld:lx="17" esld:ly="30"/>
> 	<esld:IEDName esld:name="XAT_242_MU2" esld:x="19" esld:y="22" esld:lx="20" esld:ly="23"/>
> 	<esld:IEDName esld:name="XAT_242_MU1" esld:x="28" esld:y="22" esld:lx="29" esld:ly="23"/>
> 	<esld:IEDName esld:name="XAT_242_C1" esld:x="40" esld:y="21" esld:lx="41" esld:ly="22"/>
> 	<esld:IEDName esld:name="XAT_232_P2" esld:x="1" esld:y="15" esld:lx="2" esld:ly="16"/>
> 	<esld:IEDName esld:name="XAT_232_C1" esld:x="47" esld:y="12" esld:lx="48" esld:ly="13"/>
> 	<esld:IEDName esld:name="XAT_2242_MU2" esld:x="1" esld:y="2" esld:lx="2" esld:ly="3"/>
> 	<esld:IEDName esld:name="XAT_2242_MU1" esld:x="61" esld:y="2" esld:lx="62" esld:ly="3"/>
> 	<esld:IEDName esld:name="XAT_2152_MU2" esld:x="61" esld:y="14" esld:lx="62" esld:ly="15"/>
> 	<esld:IEDName esld:name="XAT_2152_MU1" esld:x="61" esld:y="6" esld:lx="62" esld:ly="7"/>
> 	<esld:IEDName esld:name="XAT_T2_P2" esld:x="45" esld:y="16" esld:lx="46" esld:ly="17"/>
> 	<esld:IEDName esld:name="XAT_T1_P2" esld:x="30" esld:y="5" esld:lx="31" esld:ly="6"/>
> 	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="24" esld:y="4" esld:lx="25" esld:ly="5"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="27" esld:y="2" esld:lx="29.5" esld:ly="5"/>
> 	<esld:IEDName esld:name="XAT_242_C1" esld:x="42" esld:y="19" esld:lx="43" esld:ly="20"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="22" esld:y="2" esld:lx="20.5" esld:ly="2"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="27" esld:y="2" esld:lx="25.5" esld:ly="2"/>
> 	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="23" esld:y="6" esld:lx="22" esld:ly="6"/>
> 	<esld:IEDName esld:name="XAT_BusA_P1" esld:x="24" esld:y="4" esld:lx="25" esld:ly="5"/>
> 	<esld:IEDName esld:name="XAT_T1_P2" esld:x="31" esld:y="5" esld:lx="32" esld:ly="6"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="40" esld:y="1" esld:lx="49" esld:ly="2.5"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="35" esld:y="3" esld:lx="36" esld:ly="4"/>
> 	<esld:IEDName esld:name="XAT_CTRL_2" esld:x="23" esld:y="3" esld:lx="24" esld:ly="4"/>
> 	<esld:IEDName esld:name="XAT_T2_P1" esld:x="14" esld:y="2" esld:lx="15" esld:ly="3"/>
> </Private>
>
> Any chance of having an edit IED menu added?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenEnergyTools/oscd-designer/pull/2#issuecomment-2415930463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFEXXZVM25SFSYTM7RGDC3Z3YHOFAVCNFSM6AAAAABPAK3ESCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMJVHEZTANBWGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

jarradraumati commented:
This is all I get now (I'm pretty sure it was there before):

![image](https://github.com/user-attachments/assets/cf651f48-ed55-40eb-9ec7-a7fb01d4c19d)

Middle click does work, but only if done on the text (not the IED symbol)

jarradraumati commented:
I take back the issue about the symbol. It looks like the edit behaviour when middle-clicking is on text only.

danyill commented:
Thanks @jarradraumati for the feedback.

I've restored the "Edit" option for the IED.

I've also made undo/redo work correctly, _but_... You have to do two undo actions when an IED moves between containers (Substation / Voltage Level / Bay).

If you have time to re-test I'd be grateful. In particular, to test the following:

* Can you find a way to get an IED in the wrong container? This includes moving between two separate Substations or containers within different Substations.
* If an IEDName is removed it should always remove the IED if it exists. I think it works correctly...
* If an IED is separately removed then there should be a menu option when placing IEDs to remove the unlinked IEDs. 
![image](https://github.com/user-attachments/assets/aa1dbaec-8f51-4181-9efb-9154e6497804). Can you test an example?

If these all work, then apart from tests I think we're done.

jarradraumati commented:
> Can you find a way to get an IED in the wrong container? This includes moving between two separate Substations or containers within different Substations.

I have not been able to do this.

> If an IEDName is removed it should always remove the IED if it exists. I think it works correctly...

I believe this is true. Deleting the IED appears to delete it from the SCD file.
However, when I delete the last IED from a container, the private element `<Private type="OpenSCD-Linked-IEDs"/>` is not cleaned up. When the last IED is moved outside of the container the private element is removed/cleaned up as I expect.

> If an IED is separately removed then there should be a menu option when placing IEDs to remove the unlinked IEDs.

Hmm there isn't an editor to delete IEDs in the demo.
I manually removed an IED from the SCD file and a menu option appeared to remove the IED from the SLD in the dropdown.
It would be nice if missing IEDs were shaded a different colour/greyed out if the IED that is on the SLD is missing from the SCD file. But this is more of a feature request than testing the functionality.

Moving an IED label outside of the voltage level it is in has an interesting outcome.
It is hidden behind it, and it will let you place it there.
It will not let you place the label in a different substation (which I think is correct).

https://github.com/user-attachments/assets/72d577ec-71b8-4c30-a776-83b8043b859b

danyill commented:
> However, when I delete the last IED from a container, the private element <Private type="OpenSCD-Linked-IEDs"/> is not cleaned up. When the last IED is moved outside of the container the private element is removed/cleaned up as I expect.

I've made sure this is cleaned up now.

> It would be nice if missing IEDs were shaded a different colour/greyed out if the IED that is on the SLD is missing from the SCD file. But this is more of a feature request than testing the functionality.

I implemented this initially, it was one of the reasons for the performance slow-down so I've left it for later on.
There's likely a better way to do this but it seems it would be a larger refactor than I'd like to do for now.

> Moving an IED label outside of the voltage level it is in has an interesting outcome.
> It is hidden behind it, and it will let you place it there.
> It will not let you place the label in a different substation (which I think is correct).

That's interesting, I've fixed it.

danyill commented:
Thanks for the testing @jarradraumati :love_letter:

danyill commented:
@JakobVogelsang I think this is ready for review (I don't have permissions to add reviewers) and I would be most grateful if you could look at over the next few weeks.

Once we agree here,  I'll do a PR for the Communication Editor to allow IED placement within that. Perhaps we should not merge this until that is complete as these plugins are somewhat dependent on each other.

It still doesn't have tests but I think it's an improvement on the previous prototype.

Incidentally I understand DIGIS 5 v 9.80 works well with the previous version also (preserves namespaces within Siemens IEDs but other tools still don't).

JakobVogelsang commented:
I have done a first functional review and have found two issues:
1. The speed of impacted when a bay or voltage level is moved tbh by a lot. I want to understand what the cause is when I get into the code review.
2. When disabling the display of IED the label of the IED still appears.

JakobVogelsang commented:
I will get familiar with the code tomorrow and try to give you a review by the end of the weekend.

danyill commented:
> I have done a first functional review and have found two issues:

Thanks! :love_letter: 

> 1. The speed of impacted when a bay or voltage level is moved tbh by a lot. I want to understand what the cause is when I get into the code review.

I don't see this -- It used to be due to expensive queries with Text elements on the IEDs but I've removed that feature on advice from Christian.

https://github.com/user-attachments/assets/78fd253d-5c66-4fe5-83f1-ab6aaced39e7

My computer is old and slow so on a substation of this size I would have guessed performance was acceptable.

Here is an example file for comparison:

[XAT_Subsystem_Testing_old_coords_removed.scd.zip](https://github.com/user-attachments/files/17840514/XAT_Subsystem_Testing_old_coords_removed.scd.zip)

> 2. When disabling the display of IED the label of the IED still appears.

https://github.com/user-attachments/assets/88aeed28-c533-4921-97a2-7624afe26877

I also don't see this, are you up-to-date on the branch?

Might be worth checking you are exactly on this branch: `issue-92-solution-substation` and _not_ `issue-92-position-ieds` which was an earlier effort.

danyill commented:
> It would be nice if missing IEDs were shaded a different colour/greyed out if the IED that is on the SLD is missing from the SCD file. But this is more of a feature request than testing the functionality.

It occurred to me that as long as we don't need this while placing / moving it can probably be reasonably performant.

I've made the label text red if the IED is not present.

![image](https://github.com/user-attachments/assets/62daa680-32e7-4150-bd01-b0f0efb01c92)

Let's see what Jakob thinks! :wink:

danyill commented:
I've deployed this to https://transpower-nz.github.io/open-scd/ so hopefully we'll find its rough edges there.

ca-d commented:
I'm utterly overjoyed playing around with this feature. It looks 100% polished and well thought-out, and integrates so well with existing UI around it that I had not a single surprise in trying to use all the facets of this complex added feature. I'd love to have this in the OET distribution, thank you so much for this painstaking work! :gift_heart:

danyill committed:
Update scl-lib

JakobVogelsang reviewed (approved):
LGTM

JakobVogelsang commented:
I like the feature as well and would like to merge it. Do you have any idea why so many tests are failing? @danyill

danyill commented:
No but I can take a little look this weekend

danyill commented:
> I like the feature as well and would like to merge it. Do you have any idea why so many tests are failing? @danyill

I think part of the answer to this is that towards the end of the initial development we asked Christian to focus on features rather than tests because we were running short of resources.

After bisecting, test failures begin at 0ac9d1c all of the test coordinates needed to be updated - I think this is because we wanted objects moved to be relative to the cursor position where it started at (0,0) whereas previously we moved the top left corner of the object moved to where the mouse cursor was which was less intuitive for the user.

This looks like it requires quite a lot of manual updates to the tests and in some cases will cause incorrect test behaviour because the "movement" on the UI is now different.

Some of the test failures surprise me as after placing a bay, `this.placing` has not been reset and I can't  presently see why  this is (what in the preceding commit causes this change in functionality):

```
 ❌ Designer > given a bay > renames reparented bays if necessary
      AssertionError: expected <oscd-designer docname="testDoc"></oscd-designer> to have property 'placing' of undefined, but got <Bay xmlns="http://www.iec.ch/61850/2003/SCL" name="B1" xmlns:esld="https://transpower.co.nz/SCL/SSD/SLD/v0" esld:x="2" esld:y="2" esld:w="3" esld:h="3" esld:lx="2" esld:ly="2">
              <ConnectivityNode name="L1" pathName="S1/V1/B1/L1"/>
            </Bay>
        at qy.<anonymous> (oscd-designer.spec.ts:630:30)

 ❌ Designer > given a bay > updates reparented bays' connectivity node paths
      AssertionError: expected <oscd-designer docname="testDoc"></oscd-designer> to have property 'placing' of undefined, but got <Bay xmlns="http://www.iec.ch/61850/2003/SCL" name="B1" xmlns:esld="https://transpower.co.nz/SCL/SSD/SLD/v0" esld:x="2" esld:y="2" esld:w="3" esld:h="3" esld:lx="2" esld:ly="2">
              <ConnectivityNode name="L1" pathName="S1/V1/B1/L1"/>
            </Bay>
        at qy.<anonymous> (oscd-designer.spec.ts:653:30)
```

I'd be happy to look at this some more but I'd likely need a few more hints.

danyill commented:
> Some of the test failures surprise me as after placing a bay, this.placing has not been reset and I can't presently see why this is (what in the preceding commit causes this change in functionality):

When I track this in my browser window outside of the test suite I can see `element.reset()` being called but when I log console statements around it in the tests I can't see it. Alas the reasons are currently beyond me :disappointed: 

All help gratefully received.

ca-d commented:
As I've stated before in Zulip, one part of the changes needed for green tests is that all mouse interactions need their y coordinates increased by a constant amount due to the pallette being repositioned to the top. Don't know if more change than that are needed.

ca-d commented:
We'll triage test coverage alongside other outstanding work on both `oscd-designer` and `oscd-diff` come April and try to get to all of these in the coming months.

danyill commented:
I guess there have been many good upstream changes however as a result this is no longer mergeable :disappointed:

danyill commented:
I shall revive this via a PR at https://github.com/OMICRONEnergyOSS/oscd-editor-sld
